### PR TITLE
Generate readfirstlane efficiently for uniform case

### PR DESCRIPTION
**Pull request by xuechen417**
The Vulkan Enviroment for SPIR-V (section "Validation Rules within a
Module",  version 1.2.140) says that:
"If an instruction loads from or stores to a resource (including atomics
and image instructions)
and the resource descriptor being accessed is not dynamically uniform,
then the operand
corresponding to that resource (e.g. the pointer or sampled image
operand) must be decorated
with NonUniform."
The usage of NonUniform decoration is modified according to spec both in
CTS and glslang.
Example:
layout(set = 0, binding = 0) uniform texture2D uTex[];
layout(set = 1, binding = 0) uniform sampler uSamp;
layout(location = 0) flat in int Index;
layout(location = 0) out vec4 FragColor;
void main()
{
// Old usage not according spec
FragColor = texture(sampler2D(uTex[nonuniformEXT(Index)], Samp),
vec2(0.5));
// New usage according spec
FragColor = texture(
nonuniformEXT(sampler2D(uTex[Index], uSamp)), ec2(0.5));
}

it produces the following SPIR-V (only extract provided)
// Old usage
...
OpDecorate %18 NonUniform
OpDecorate %21 NonUniform
...
%5 = OpLabel
%17 = OpLoad %int %Index
%18 = OpCopyObject %int %17 <-- The index is NonUniform
%20 = OpAccessChain %_ptr_UniformConstant_10 %uTex %18
%21 = OpLoad %10 %20 <-- The image itself is NonUniform
%25 = OpLoad %22 %uSamp
%27 = OpSampledImage %26 %21 %25
%31 = OpImageSampleImplicitLod %v4float %27 %30 <-- The sampled image is
not NonUniform
...
// New usage
...
OpDecorate %33 NonUniform
...
%10 = OpTypeImage %6 2D 0 0 0 1 Unknown
...
%31 = OpTypeSampledImage %10
...
%32 = OpSampledImage %31 %20 %30
%33 = OpCopyObject %31 %32
%37 = OpImageSampleImplicitLod %7 %33 %36
...

Fixes:
- https://github.com/GPUOpen-Drivers/llpc/issues/840
- dEQP-VK.descriptor_indexing.sampled_image*.

The reason of cts failutres is that the descriptor index is mistook as
uniform in indexDescPtr which it should be non-uniform.
Solution:
- Remove the non-uniform flag in indexDescPtr, which can make the test
pass.
- Add two flags ImageFlagEnforceReadFirstLaneImage/Sampler. Mark the flag
if the accessing image or sampled image array isn't decorated with
nonuniform.
- Enforce readfirstlane on the resource descriptor if they are marked as
ImageFlagEnforceReadFirstLaneImage/Sampler.
- Add a otpimization pass in middle-end to push up readfirstlane to
transform the use of vgpr to sgpr  as many as possible.

### CODE REVIEW

**Comment by xuechen417** [comment]
This PR is based on the reverted PR #976 with a modification of the condition of enforcing readfirstlane to avoid game corruption. 
The original condition is "The resource descriptor without nonuniform decoration will be enforced readfirstlane".
The modified condition is "The resource descriptor in an array without nonuniform decoration will be enforced readfirstlane".

**Comment by amdvlk-admin** [comment]
### Test summary for commit 000cc5e7aef37767c5bf5aef34beba1ce4e84dc5
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 8abed125acda77ccf700ddd0f5bf0114d25ac135</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 2f28502c5e6e4f059a014ab30001a191328f7a5d</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 4ad346045fdd6ef7593ce198b78ed563b74b68c5</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/224557)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 26814/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29306/56120 (52.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26814/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29306/56120 (52.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 25403/56197 (45.2%)</li>
<li>Failed: 0/56197 (0.0%)</li>
<li>Skipped: 30794/56197 (54.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 26829/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29291/56120 (52.2%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -5292,10 +5303,27 @@ void SPIRVToLLVM::getImageDesc(SPIRVValue *bImageInst, ExtractedImageInfo *info)
...
+      if (sampler->getOpCode() == OpLoad)
+        samplerLoadSrc = static_cast<SPIRVLoad *>(sampler)->getSrc();
     }
     scanBackInst = static_cast<SPIRVInstTemplateBase *>(scanBackInst)->getOpValue(0);
     if (scanBackInst->hasDecorate(DecorationNonUniformEXT))
       info->flags |= lgc::Builder::ImageFlagNonUniformImage;
+    if (scanBackInst->getOpCode() == OpLoad)
+      imageLoadSrc = static_cast<SPIRVLoad *>(scanBackInst)->getSrc();
+  }
+  // Set enforce readfristlane flag for accessing image or sampled image array
```
Comment (dstutt) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:5315]: Spelling

**Review by dstutt — APPROVED** [review]
So - with the new glsl etc - we only get the nonUniform decoration on the Image or Sampler (and glsl handles all of that when emitting the SPIR-V)?

Make sense to me.

Minor nits: a couple of spelling mistakes (approving change though)

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -5259,29 +5252,47 @@ void SPIRVToLLVM::getImageDesc(SPIRVValue *bImageInst, ExtractedImageInfo *info)
...
+      imageAccessChain = bImagePtr;
       bImagePtr = operands[0];
     }
     assert(bImagePtr->getOpCode() == OpVariable || bImagePtr->getOpCode() == OpFunctionParameter);
     if (bImageInst->hasDecorate(DecorationCoherent))
       info->flags |= lgc::Builder::ImageFlagCoherent;
     if (bImageInst->hasDecorate(DecorationVolatile))
       info->flags |= lgc::Builder::ImageFlagVolatile;
+
+    // Set enforce readfristlane flag for accessing image array
```
[the remark this answers is not in the document]
Comment (xuechen417) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:5271]: Fixed

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 1 above)
Comment (xuechen417) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:5315]: Fixed

**Review by xuechen417 — COMMENTED** [review]
V1: Fix spelling.  @dstutt new glsl puts nonuniform decoration outsided the sampled image contructor. For pure image or sampler array, the nonuniform still decorates the index.

**Pull request merged**